Implement debug HTTP APIs (#63)

rdmitr opened the pull request:
Fixes https://github.com/getAlby/nostr-wallet-connect-next/issues/63
- [x] add wails support

rolznz commented on models/lnclient/lnclient.go:
```diff
@@ -57,6 +57,12 @@ type LNClient interface {
 	GetOnchainBalance(ctx context.Context) (int64, error)
 }
 
+type DebugClient interface {
```
I think these should actually be added to the standard LNClient and just be treated as normal methods

rolznz commented on http_service.go:
```diff
@@ -74,6 +75,10 @@ func (httpSvc *HttpService) RegisterSharedRoutes(e *echo.Echo) {
 
+	e.POST("/api/debug/send-payment-probes", httpSvc.sendPaymentProbesHandler, authMiddleware)
+	e.POST("/api/debug/send-spontaneous-payment-probes", httpSvc.sendSpontaneousPaymentProbesHandler, authMiddleware)
+	e.GET("/api/debug/peers", httpSvc.listPeers, authMiddleware)
```
I don't think we need to add the /debug prefix

rolznz commented:
@rdmitr are you able to do a simple frontend page that can call these methods?

rdmitr commented:
> @rdmitr are you able to do a simple frontend page that can call these methods?

My frontend skills are close to zero to be honest. I can do it if necessary but someone with the right skills will be able to handle it much faster ;)

rolznz commented:
@rdmitr are you also able to add the methods to request log output?

rdmitr commented:
> @rdmitr are you also able to add the methods to request log output?

@rolznz yep, done but not tested

rolznz commented:
@pavanjoshi914 once your extension work is done, could you help work on the frontend? we would need a new "debug" page to call these API calls - if any parameters are needed they can be requested with a simple window.prompt - and then log the output as simply as possible (maybe even just to the window console logs)

rolznz commented on models/api/dbgapi.go:
```diff
@@ -0,0 +1,26 @@
+package api
+
+type SendPaymentProbesRequest struct {
```
Can these be moved to the normal api models?

pavanjoshi914 replied:
resolved

rolznz commented:
@rdmitr I think we only have log output for the LDK node right? can we have one endpoint for the node logs (the current one) and another endpoint for the app logs?

rdmitr commented:
@rolznz @pavanjoshi914 

I have added another endpoint, `/api/get-app-log-output`, to read the app log output. It requires log source to be specified, `general` or `error`, since the app keeps those in separate files.

I have also renamed the `api/get-log-output` endpoint to `api/get-ln-log-output` to make it more clear. The frontend file is updated as well.

pavanjoshi914 commented:
@rdmitr can you please try and see why the checks are failing

rolznz commented:
@pavanjoshi914 I think it was just a random error. I re-ran the workflow and it passed

rolznz commented on http_service.go:
```diff
@@ -95,6 +97,11 @@ func (httpSvc *HttpService) RegisterSharedRoutes(e *echo.Echo) {
 
+	e.POST("/api/send-payment-probes", httpSvc.sendPaymentProbesHandler, authMiddleware)
+	e.POST("/api/send-spontaneous-payment-probes", httpSvc.sendSpontaneousPaymentProbesHandler, authMiddleware)
+	e.POST("/api/get-ln-log-output", httpSvc.getLnLogOutput, authMiddleware)
```
suggestion: make the log endpoints `GET`s to be restful. Maybe GET `/logs` and we can use query params like `type=app` or `type=node` and `lines=1000`

rdmitr replied:
Right, I sort of copy-pasted this. Will fix shortly

rolznz commented:
@pavanjoshi914 can you resolve the conflicts?

rolznz commented on models/api/api.go:
```diff
@@ -132,6 +132,46 @@ type NewOnchainAddressResponse struct {
+
+type GetAppLogOutputRequest struct {
+	MaxLen int    `json:"maxLen"`
+	Source string `json:"source"`
```
maybe it would be easier if we just merge the logs and keep the requests the same for app vs node logs.

Also, can we use log lines instead of characters for the max length?

rdmitr replied:
> maybe it would be easier if we just merge the logs and keep the requests the same for app vs node logs

In fact, I find it a bit odd that we keep separate log files for errors and everything else :) Would logging everything to the same file be more manageable?

rolznz replied:
@bumi I agree, can we change this?

bumi replied:
totally. when I had looked at logging I just wanted to log to a file AND to stdout.

rolznz replied:
So `log/nwc-general.log` and `log/nwc-error.log` would just become `log/nwc.log` (with file rotation - I think `MaxAge` would be 3 days and `MaxBackups` would be also be 3 - one per day if I understand this correctly).

rolznz commented on utils.go:
```diff
@@ -25,3 +32,49 @@ func GetStartOfBudget(budget_type string, createdAt time.Time) time.Time {
 	}
 }
+
+func LoggedClose(logger *logrus.Logger, closer io.Closer) {
```
What is this for?

rdmitr replied:
When we do a plain `defer f.Close()`, we lose the error potentially returned by `Close()`, and linters complain about this. This is a simple wrapper to catch and log those errors, if they happen.

rolznz replied:
Hmm, but it doesn't seem to be used

rdmitr replied:
Right, I have changed the code quite a lot 🤦‍♂️  Will remove it now

rdmitr commented:
I have merged the two log endpoints into a single one:

`GET /api/log/app?source=error&maxLen=65536`
`GET /api/log/node?maxLen=65536`

rolznz commented:
@pavanjoshi914 could you look at the frontend parts again?

rdmitr commented:
I have changed the logging configuration to use a single app log file, nwc.log

pavanjoshi914 commented on http_service.go:
```diff
@@ -96,6 +98,10 @@ func (httpSvc *HttpService) RegisterSharedRoutes(e *echo.Echo) {
 
+	e.POST("/api/send-payment-probes", httpSvc.sendPaymentProbesHandler, authMiddleware)
+	e.POST("/api/send-spontaneous-payment-probes", httpSvc.sendSpontaneousPaymentProbesHandler, authMiddleware)
+	e.GET("/api/log/:type", httpSvc.getLogOutput, authMiddleware)
```
@rdmitr why GET here? we need to pass maxLen in request body right?

rolznz commented:
@rdmitr for sending spontaneous payment probes with an amount and node id does not work for me. I sent to Amboss' node (03006fcf3312dae8d068ea297f58e2bd00ec1ffe214b793eda46966b6294a53ce6) from their keysend bilboard https://amboss.space/

and I get ```
API Response: {
  "error": "NodeError: InvalidPublicKey: The given public key is invalid."
}
```

Note: I can keysend to this node fine. Can you have a look and see if there's an issue with the bindings?

rolznz commented on frontend/src/components/layouts/AppLayout.tsx:
```diff
@@ -111,6 +112,10 @@ export default function AppLayout() {
           <ShieldCheck className="h-4 w-4" />
           Permissions
         </MenuItem>
+        <MenuItem to="/advanced">
```
Can we move this into the settings page nav and call it debug tools? this is way too obvious where it is now

pavanjoshi914 replied:
resolved

rolznz commented on frontend/src/screens/Advanced/Debug.tsx:
```diff
@@ -0,0 +1,101 @@
+        </Button>
+        <Button
+          onClick={() => {
+            let maxLen = window.prompt("Enter max length:");
```
```suggestion
            let maxLen = window.prompt("Enter max length (in characters):");
```

rolznz commented on frontend/src/screens/Advanced/Debug.tsx:
```diff
@@ -0,0 +1,101 @@
+        </Button>
+        <Button
+          onClick={() => {
+            let maxLen = window.prompt("Enter max length:");
```
can you use `const` where you do not re-assign variables? do you have linting in your IDE?

pavanjoshi914 replied:
true! resolved

rolznz commented on frontend/src/screens/Advanced/Debug.tsx:
```diff
@@ -0,0 +1,101 @@
+            let maxLen = window.prompt("Enter max length:");
+
+            if (maxLen)
+              apiRequest(`/api/log/app?source=error&maxLen=${maxLen}`, "GET");
```
`source=error` should be removed now, right?

pavanjoshi914 replied:
hmmm! i was getting error before without it. did we removed this recently? resolved

rolznz commented on frontend/src/screens/Advanced/Debug.tsx:
```diff
@@ -0,0 +1,101 @@
+              });
+          }}
+        >
+          Send Spontaneous Payment Probes
```
![image](https://github.com/getAlby/nostr-wallet-connect-next/assets/33993199/44d3c2eb-200d-4536-b18b-9a2579ef5192)

@pavanjoshi914 These buttons don't look that great, especially where the text is too long.

What about "Probe Invoice" and "Probe Keysend"

And can the keysend one be in sats and multiply the value by 1000 before sending it to the backend? I don't think anyone will probe in millisats

pavanjoshi914 replied:
resolved

rolznz commented on http_service.go:
```diff
@@ -580,3 +597,81 @@ func (httpSvc *HttpService) setupHandler(c echo.Context) error {
 }
+
+func (httpSvc *HttpService) sendPaymentProbesHandler(c echo.Context) error {
+	var sendPaymentProbesRequest api.SendPaymentProbesRequest
```
@rdmitr can you please move the actual logic in these methods to `api.go` so that they can also be called by the Wails router? (we support running the app in two modes - http or wails)

rolznz replied:
Can you also make sure to check the lnClient is not nil at the top of each method (like the existing api methods do)

rdmitr replied:
Done, logic has been moved, and Wails router has been updated 👍

rolznz commented:
@rdmitr could you look at adding the same methods too the wails router? note: it's a custom router and not as nice as the HTTP one - if you have any ideas of a way to improve it let me know

rdmitr commented:
> @rdmitr could you look at adding the same methods too the wails router? note: it's a custom router and not as nice as the HTTP one - if you have any ideas of a way to improve it let me know

Yep, methods added. I could not come up with ideas for improving the router, but I'll keep it in mind

rolznz commented:
> @rolznz the public key you mention parses just fine in LDK. However, in Debug.tsx the request body seems to use snake_case instead of camelCase for request field names:

@pavanjoshi914 can you check this? please test the buttons/code you add. If you don't know how to test it let me know

rolznz commented:
> Could it be the reason? Should I rename fields to snake case? I made them camel case for consistency with other request structures in the same file (api.go).

@rdmitr I think it's good to stay consistent. I think this was not tested on the frontend. Sorry about that.

pavanjoshi914 commented:
@rolznz wouldn't it be consistent to use same param names as defined in function definitions. i just navigated to function definitions(which use snake case) instead of looking at API handler functions and models and used them in frontend.

```func (gs *LDKService) SendSpontaneousPaymentProbes(ctx context.Context, amount_msat uint64, node_id string)```

uses snake case. but from the handlers we need to pass args as camelcase?

rolznz commented:
@pavanjoshi914 thanks for raising this. There are a few different domains here which could have their own conventions: the JS code, the JSON (payloads) and the Go code. But actually our current convention is to use camelCase in all these places. In the rust code (LDK node) the convention is different and uses snake_case.

So there are some places in this code that should use camelCase where it is currently snake_case.

rolznz commented on http_service.go:
```diff
@@ -339,6 +345,17 @@ func (httpSvc *HttpService) mempoolLightningNodeHandler(c echo.Context) error {
 }
 
+func (httpSvc *HttpService) listPeers(c echo.Context) error {
+	peers, err := httpSvc.svc.lnClient.ListPeers(httpSvc.svc.ctx)
```
@rdmitr can you update this to go through the API?

rolznz replied:
And also add it to the wails router

rdmitr replied:
Yep, done!

rolznz commented on http_service.go:
```diff
@@ -580,3 +597,64 @@ func (httpSvc *HttpService) setupHandler(c echo.Context) error {
+	}
+
+	logType := c.Param("type")
+	if logType != api.LogTypeNode && logType != api.LogTypeApp {
```
we have this same check in 3 places that we have to maintain, maybe I would just have this check in the API?

rdmitr replied:
I feel that this kind of logic (validating input models) belongs to the controller domain; the API doesn't really have a check per se, just a generic `else` clause when checking `logType`.

If we leave this check in API only, we will need to introduce a new error type, "validation error", in the API domain, so that the request handler methods that call the API methods can discern it from the generic "something went south" error and respond appropriately, i.e. with HTTP codes 40x (client messed up) vs 50x (we messed up). So it felt more natural to do that in the handlers 🤔

rolznz replied:
@rdmitr good point, thanks!

rdmitr commented:
@pavanjoshi914 actually, the method you mention, `SendSpontaneousPayment()`, should use camelCase for args. Thanks for mentioning this — I shall fix it too. It's not a breaking change and it won't affect the frontend at all.

rolznz reviewed (approved):
tACK
